File: lib/747.largest_number_at_least_twice_of_others.dart
void main() {
  print(dominantIndex([3, 6, 1, 0]));
  print(dominantIndex([1, 2, 3, 4]));
  print(dominantIndex([0, 0, 3, 2]));
  print(dominantIndex([3, 0, 0, 2]));
}

int dominantIndex(List<int> nums) {
  if (nums.isEmpty) {
    return -1;
  }

  var maxValueIndex = 0;
  var secondMaxValueIndex = -1;

  for (int i = 1; i < nums.length; i++) {
    if (nums[i] > nums[maxValueIndex]) {
      secondMaxValueIndex = maxValueIndex;
      maxValueIndex = i;
    } else if (secondMaxValueIndex == -1 || nums[i] > nums[secondMaxValueIndex]) {
      secondMaxValueIndex = i;
    }
  }

  return secondMaxValueIndex == -1 || (nums[maxValueIndex] / 2) >= nums[secondMaxValueIndex]
      ? maxValueIndex
      : -1;
}

File: src/747.largest_number_at_least_twice_of_others.kt
fun main() {
    println(dominantIndex(intArrayOf(3, 6, 1, 0)))
    println(dominantIndex(intArrayOf(1, 2, 3, 4)))
    println(dominantIndex(intArrayOf(0, 0, 3, 2)))
    println(dominantIndex(intArrayOf(3, 0, 0, 2)))
}

fun dominantIndex(nums: IntArray): Int {
    if (nums.isEmpty()) {
        return -1
    } else if (nums.size == 1) {
        return 0;
    }

    var maxValueIndex = 0
    var secondValueIndex = -1

    for (i in 1..<nums.size) {
        if (nums[i] > nums[maxValueIndex]) {
            secondValueIndex = maxValueIndex
            maxValueIndex = i
        } else if (secondValueIndex == -1 || nums[i] > nums[secondValueIndex]) {
            secondValueIndex = i
        }
    }

    return if (secondValueIndex == -1 || nums[maxValueIndex] / 2 >= nums[secondValueIndex]) {
        maxValueIndex
    } else {
        -1
    }
}